config/main.py

```diff
@@ -673,8 +684,11 @@ def _stop_services():
         pass
 
     click.echo("Stopping SONiC target ...")
-    clicommon.run_command("sudo systemctl stop sonic.target")
-
+    try:
+        clicommon.run_command("sudo systemctl stop sonic.target")
```

Comment:
Instead of catching "Job Cancellation", we are also handling "Job failure" here.

Since, we're just dealing with Job Cancellation here, How about localizing the exception catch mechanism to just the "Cancellation".

Reply:
The intention is indeed to handle all possible failures when we are stopping services. The "config reload" operation is a system affecting event. We need to ensure that this command performs the intended action  in the entirety and also brings the system back to a well defined state. By not handling errors, we may be leaving the system in a broken state and in some cases inaccessible.

The proposed changes handle errors in the stop flow. It is to be noted that we are restarting all services at a later stage even if failures were seen during stop.

It is clear that the "config reload" is not robust enough. ZTP relies on "config reload" to load user provided configuration and as well as its own ZTP default configuration. So it is essential we fix "config reload" and not depend on ZTP to perform config reload again. Having said that, ZTP does perform a "config reload <ztp-default-config>" if the "config reload" pushed by the user fails. This is to ensure that ZTP retries a new configuration that user might push after noticing the errors.

Reply:
I agree, we are restarting these services later in the pipeline but from the current design, we are trying to stop all the dependent services before applying the configuration.

I am not entirely sure about the repercussions of keeping the services active when the config changes are applied. If that is deemed safe, we can proceed with moving forward with the config reload, even if all the services haven't stopped.

In any case, i have a suggestion to reduce the chances of services being in active state

How about having a fixed number of retries on failed services (maybe 3) and if the stop job doesn't success after 3 tries, we can then choose to proceed with whatever approach we choose to adopt (Exit or Move Forward)

```
MAX_TRIES = 3
services_to_stop = _get_sonic_services()
failed_execution  = []
tries  = 0
while tries < MAX_TRIES:
        for s in services_to_stop:
                try:
                    clicommon.run_command("systemctl stop {}".format(s))
                except SystemExit as err:
                    failed_execution.append(s)
                    clicommon.run_command("systemctl reset-failed {}".format(s))
      tries  += 1   
      if tries >= MAX_TRIES and failed_execution:
         click.echo("Error encountered while stopping these services {}. Continuing..".format(failed_execution))
      elif failed_execution:
         services_to_stop = failed_execution
         failed_execution = []
     else:
         break
```
Let me know if this approach has any setbacks.

Reply:
This issue was present even before sonic.target was introduced. We used to see this issue with swss.target and its failure was handled. However, after sonic.target we are back to seeing the stop operation being canceled. From my experience with SONiC and 'config reload', there is no explicit requirement to retry stopping failed individual services. One may argue why 3 and why not something else. I would thoroughly test the current approach before making an further assumptions on possibilities. But that is just my personal choice.

Reply:
If the current solution is robust enough to work around the race conditions, then i guess it's okay to stick with this.